### PR TITLE
[NN] Add EGNN & PNA

### PR DESCRIPTION
**Pull request by rudongyu**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:1]: "Convolution" -> "Convolutional" to be consistent with the paper

**Comment by mufeili** [comment]
Lint issues reported by CI

"""
[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-277)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-278)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-279)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-280)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-281)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-282)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-283)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-284)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-285)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-286)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-287)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-288)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-289)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-290)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-291)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-292)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-293)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-294)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-295)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-296)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-297)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-298)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-299)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-300)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-301)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-302)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-303)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-304)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-305)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-306)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-307)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-308)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-309)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-310)[](https://ci.dgl.ai/blue/organizations/jenkins/dgl/detail/PR-3901/1/pipeline#step-101-log-311)python/dgl/nn/pytorch/conv/pnaconv.py:65:0: C0301: Line too long (101/100) (line-too-long)

python/dgl/nn/pytorch/conv/pnaconv.py:73:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:78:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:85:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:92:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:95:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:101:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:108:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:110:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:140:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:141:50: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:157:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:185:0: C0301: Line too long (115/100) (line-too-long)

python/dgl/nn/pytorch/conv/pnaconv.py:189:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:195:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:203:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:208:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:253:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/pnaconv.py:7:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:10:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:13:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:16:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:19:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:22:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:35:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:38:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:41:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:44:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:44:22: W0613: Unused argument 'D' (unused-argument)

python/dgl/nn/pytorch/conv/pnaconv.py:44:30: W0613: Unused argument 'delta' (unused-argument)

python/dgl/nn/pytorch/conv/pnaconv.py:47:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:50:0: C0116: Missing function or method docstring (missing-function-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:64:0: C0115: Missing class docstring (missing-class-docstring)

python/dgl/nn/pytorch/conv/pnaconv.py:96:4: C0116: Missing function or method docstring (missing-function-docstring)

************* Module dgl.nn.pytorch.conv.egnnconv

python/dgl/nn/pytorch/conv/egnnconv.py:141:0: C0303: Trailing whitespace (trailing-whitespace)
"""

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:7]: "E(n) Equivariant" -> "Equivariant Graph Convolutional Layer"?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:26]: :math:`\phi_e, \phi_x, \phi_h`?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:54]: "|x_i - x_j|^2" -> "||x_i - x_j||^2"?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:115]: "The coordinate feature of shape :math:`(N, *)`" -> "The coordinate feature of shape :math:`(N, *)`."

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:116]: Can `*` be an n-tuple rather than a positive integer?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:137]: `edge_feat` should be provided as long as `self.edge_feat_size > 0`.

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:76]: You can compute the coordinate differences for all edges with `graph.apply_edges(fn.u_sub_v('x', 'x', 'x_diff'))` before invoking `graph.update_all`. This might be more efficient.

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:79]: L77、L79 can also be computed outside of `update_all` at once.

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:85]: It might be more efficient to pre-compute messages with a UDF for all edges and then use dgl's built-in message and reduce functions in `update_all`.

File: tests/pytorch/test_nn.py
Comment (mufeili) [inline tests/pytorch/test_nn.py:1430]: Add a case for `edge_feat_size=0`?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:12]: Should we change "\sum" to "\sum_{j\in\mathcal{N}(i)}" for the current implementation?

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/egnnconv.py:13]: Ditto

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:137]: "List" -> "list"

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:138]: "specify" -> "specifies"

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:155]: Ditto

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:174]: Add an empty line above.

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:176]: Add an empty line above.

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:182]: `('mean', 'max', 'sum')` is a tuple while the parameter description requires a list. Same for `('identity', 'amplification')`.

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:222]: Is there a description of it in the paper?

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:99]: If `edge_feat_size > 0`, then it seems that `edge_feat` cannot be None.

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:252]: Is this the original implementation in PNA? I prefer having an explicit input argument for residual connection like `res=True`. If `in_size` is not equal to `out_size`, then use a linear layer to project input node features.

**Review by mufeili — CHANGES_REQUESTED** [review]
Done a first pass.

File: python/dgl/nn/pytorch/conv/egnnconv.py
Comment (jermainewang) [inline python/dgl/nn/pytorch/conv/egnnconv.py:99]: Two suggestions:

First, for message function, our convention is to make it a method of the module.
```python
class MyModule(nn.Module):
    def __init__(self):
        ...
    def message(self, edges):  # message function
        ...
    def forward(self, g, ...):
        g.update_all(self.message, ...)  # pass message method to update_all
```

Second, you should try to use built-in reducers instead of UDF reducers.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/nn/pytorch/conv/pnaconv.py
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/pnaconv.py:213]: `residual` should be an argument to `__init__`.

**Comment by mufeili** [comment]
I'm good. You can work with @jermainewang and merge the PR once you get his approval.

**Review by jermainewang — APPROVED** [review]
LGTM

**Review by mufeili — APPROVED** [review]
LGTM

**Pull request merged**
